artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/BucketMap.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.utils.collections;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+public class BucketMap<K, V> implements Map<K, V> {
+
+   private final Map<Integer, V> buckets = new IntHashMap<>();
+   private final int bucketCount;
+
+   public BucketMap(int bucketCount) {
+      this.bucketCount = bucketCount;
+   }
+
+   @Override
+   public int size() {
+      return buckets.size();
+   }
+
+   @Override
+   public boolean isEmpty() {
+      return buckets.isEmpty();
+   }
+
+   @Override
+   public boolean containsKey(Object key) {
+      return buckets.containsKey(getBucket(key));
+   }
+
+   private int getBucket(Object key) {
+      return (key.hashCode() & Integer.MAX_VALUE) % bucketCount;
```

Comment:
Why masking it with Integer.MAX_VALUE?
The % op could be replaced by an & (bucketCount-1) with power of 2 bucketCount

Reply:
So its always positive